modules/core/src/main/java/org/apache/synapse/rest/RESTRequestHandler.java

```diff
@@ -75,32 +75,24 @@ private boolean dispatchToAPI(MessageContext synCtx) {
         Collection<API> apiSet = synCtx.getEnvironment().getSynapseConfiguration().getAPIs();
         //Since swapping elements are not possible with sets, Collection is converted to a List
         List<API> defaultStrategyApiSet = new ArrayList<API>(apiSet);
-
         API defaultAPI = null;
-        for (API api : apiSet) {
-            api.setLogSetterValue();
-            if ("/".equals(api.getContext())) {
-                defaultAPI = api;
-            } else if (api.getVersionStrategy().getClass().getName().equals(DefaultStrategy.class.getName())) {
-                //APIs whose VersionStrategy is bound to an instance of DefaultStrategy, should be skipped and processed at last.
-                //Otherwise they will be always chosen to process the request without matching the version.
-                defaultStrategyApiSet.add(api);
-            } else if (api.getVersionStrategy().getClass().getName().equals(ContextVersionStrategy.class.getName())
-                    || api.getVersionStrategy().getClass().getName().equals(URLBasedVersionStrategy.class.getName())) {
+        if (null != synCtx.getProperty(RESTConstants.IS_ALREADY_PROCESSED)) {
+            API api = (API) synCtx.getProperty("PROCESSED_API");
+            identifyAPI(api, synCtx);
+        } else {
+            for (API api : apiSet) {
+                identifyAPI(api, synCtx);
+            }
+
+            for (API api : defaultStrategyApiSet) {
```

Comment:
can't we combine this and the above for loop into one loop?

Reply:
As they loop over two different sets. So we can't combine them.

Reply:
Having two for loops here is costly. We should do something here to eliminate one.

Reply:
It was the same two loops we had from the initial implementation of the RESTRequestHandler Class.